clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -3104,8 +3104,28 @@ static void handleWorkGroupSize(Sema &S, Decl *D, const ParsedAttr &AL) {
       return;
   }
 
+  ASTContext &Ctx = S.getASTContext();
+
+  if (AL.getKind() == ParsedAttr::AT_ReqdWorkGroupSize) {
+    if (const auto *A = D->getAttr<SYCLIntelNumSimdWorkItemsAttr>()) {
+      Optional<llvm::APSInt> NumSimdWorkItems =
+          A->getValue()->getIntegerConstantExpr(Ctx);
+      Optional<llvm::APSInt> XDimVal = XDimExpr->getIntegerConstantExpr(Ctx);
```

Comment:
We evaluate the constant expression potentially twice in this function (once here and once below), so it might make sense to hoist this out of the `if` loop and use the same calculation in both places.

One thing to double-check is whether `YDimVal` and `ZDimVal` can ever be null when we create the default values from `SetDefaultValue()`. I don't think they can be null within this `if` block, but it seems plausible they could be null in the `D->getAttr<WorkGroupAttr>()` block below (which could be an existing bug).

Also, what if any of `[XYZ]DimVal` is value-dependent? (This may also be an existing bug.)

Reply:
Thanks @AaronBallman for the feedback. I will check the null value issue.

Reply:
>>Also, what if any of [XYZ]DimVal is value-dependent? (This may also be an existing bug.)

addIntelSYCLTripleArgFunctionAttr() function that calls checkMaxWorkSizeAttrArguments() and checks if any of [XYZ]DimVal is value dependent,

i added this inside checkMaxWorkSizeAttrArguments() 

if (XDimExpr->isValueDependent() ||
      (YDimExpr && YDimExpr->isValueDependent()) ||
      (ZDimExpr && ZDimExpr->isValueDependent()))
    return false;

This part covers here. I do not think we need to check here again. @AaronBallman, what is your opinion?

Reply:
The X, Y, and Z dim expressions in this case are obtained from `AL` on this line: https://github.com/intel/llvm/pull/3170/files#diff-2a5bdb2d9f07f8d77de51d5403d349c22978141b6de6bd87fc5e449f5ed95becR3078 and nothing is checking whether those expressions are value-dependent before calling `getIntegerConstantExpr()` on them here. So I think this new code will cause a failed assertion on a dependent value.

Reply:
> The X, Y, and Z dim expressions in this case are obtained from `AL` on this line: https://github.com/intel/llvm/pull/3170/files#diff-2a5bdb2d9f07f8d77de51d5403d349c22978141b6de6bd87fc5e449f5ed95becR3078 and nothing is checking whether those expressions are value-dependent before calling `getIntegerConstantExpr()` on them here. So I think this new code will cause a failed assertion on a dependent value.

I have checked whether those expressions are value-dependent before calling getIntegerConstantExpr() on them here.

Reply:
>>We evaluate the constant expression potentially twice in this function (once here and once below), so it might make sense to hoist this out of the if loop and use the same calculation in both places.

Done

Reply:
> One thing to double-check is whether `YDimVal` and `ZDimVal` can ever be null when we create the default values from `SetDefaultValue()`. I don't think they can be null within this `if` block, but it seems plausible they could be null in the `D->getAttr<WorkGroupAttr>()` block below (which could be an existing bug).


`YDimVal` and `ZDimVal`  seems taking null value in  `D->getAttr<WorkGroupAttr>()` block . The following test cases fail. I have created new issue https://github.com/intel/llvm/issues/3175 to address this problem in a separate PR.

struct TRIFuncObjGood11 {
   //expected-warning@+1 {{attribute 'reqd_work_group_size' is already applied with different parameters}}
  [[intel::reqd_work_group_size(64)]] [[intel::reqd_work_group_size(128)]]void
  operator()() const {}
};

struct TRIFuncObjGood3 {
   //expected-warning@+1 {{attribute 'reqd_work_group_size' is already applied with different parameters}}
  [[intel::reqd_work_group_size(64, 64)]] [[intel::reqd_work_group_size(128, 128)]]void
  operator()() const {}
};